aspnetcore/performance/rate-limit.md

```diff
@@ -0,0 +1,186 @@
+| 0    | 100       | 20    | 0                     | 80 |
+| 10   | 80        | 30    | 0                     | 50 |
+| 20   | 50        | 40    | 0                     | 10 |
+| 30   | 10        | 30    | 20                    | 0  |
+| 40   | 0         | 10    | 30                    | 20 |
+| 50   | 20        | 10    | 40                    | 50 |
+| 60   | 50        | 35    | 30                    | 45 |
+
+The following code uses the sliding window rate limiter:
+
+:::code language="csharp" source="~/../AspNetCore.Docs.Samples/fundamentals/middleware/rate-limit/WebRateLimitAuth/Program.cs" id="snippet_slide":::
+
+<a name="token"></a>
+
+### Token bucket limiter
+
+The token bucket limiter is similar to the sliding window limiter, but rather than adding back the requests taken from the expired segment, a fixed number of tokens are added each replenishment period. The tokens added each segment can't increase the available tokens to a number higher than the token bucket limit. The following table shows a token bucket limiter with a limit of 100 tokens and a 10-second replenishment period:
+
+| Time | Available | Taken | Added | Carry over |
+| ---- | ----      | ------| ------| ---- |
+| 0    | 100       | 20    | 0     | 80 |
+| 10   | 80        | 10    | 20    | 90 |
+| 20   | 90        |  5    | 15    | 100 |
+| 30   | 100       | 30    | 20    | 90  |
+| 40   | 90        |  6    | 16    | 100 |
+| 50   | 100       | 40    | 20    | 80 |
+| 60   | 80        | 50    | 20    | 50 |
+
+The following code uses the token bucket limiter:
+
+:::code language="csharp" source="~/../AspNetCore.Docs.Samples/fundamentals/middleware/rate-limit/WebRateLimitAuth/Program.cs" id="snippet_token":::
+
+When `autoReplenishment` is set to `true`, an internal timer replenishes the tokens every `replenishmentPeriod`; when set to `false`, the app must call `TryReplenish` on the limiter.
+
+<a name="concur"></a>
+
+### Concurrency limiter
+
+The concurrency limiter limits the number concurrent requests. Each request reduces the concurrency limit by one. When a request completes, the limit is increased by one. Unlike the other requests limiters that limit the total number of requests for a specified period, the concurrency limiter limits only the number of concurrent requests and doesn't cap the number of requests in a time period.
+
+The following code uses the concurrency limiter:
+
+:::code language="csharp" source="~/../AspNetCore.Docs.Samples/fundamentals/middleware/rate-limit/WebRateLimitAuth/Program.cs" id="snippet_concur":::
+
+## Limiter algorithm comparison
+
+The fixed, sliding, and token limiter all limit the maximum number of requests in a time period. The concurrency limiter limits only the number of concurrent requests and doesn't cap the number of requests in a time period. The cost of an endpoint should be considered when selecting a limiter. The cost of an endpoint includes the resources used, for example, data access, CPU, and I/O.
```

Comment:
Is time (in the abstract _wibbly-wobbly-timey-wimey_ sense) also worth mentioning as a "cost" consideration?